providers/kind/update_rapl_permissions.sh

```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+# Base directory
+BASE_DIR="/sys/devices/virtual/powercap/intel-rapl"
+
+# Find all 'energy_uj' files in directories starting with 'intel-rapl' and change their permissions
+find $BASE_DIR -type f -name 'energy_uj' -path "$BASE_DIR/intel-rapl*" -exec sudo chmod a+r {} \;
```

Comment:
Do we really need to change the file permission. Not sure if this is a good idea....

Reply:
I also thought about that. Kepler needs access to the files "energy_uj" in the folders intel-rapl* as it decides to use a power source or not based on that. Another approach would be to run the script inside the node directly (maybe on a symlink?) after mounting but the issue is that it has to be before Kepler registers that there is no power source in power.go and thus uses prediction model because it does not have the necessary access to the files. This approach allowed me to do it without touching Keplers Code base, but maybe that would be more secure than expecting the user to change the permissions on its host. 

Do you have a suggestion on how to approach this in a more secure fashion?

Reply:
We need to debug what is happening to your system. I am also using Kind cluster on BM, but I don't need to mount and change the RAPL files permissions...

This is the the permission of the files in my host:
```
mamaral@nx12:~/go/src/github.com/marceloamaral/kepler$ ls -la /sys/class/powercap
total 0
drwxr-xr-x  2 root root 0 Sep 24 22:36 .
drwxr-xr-x 75 root root 0 Sep 13 19:19 ..
lrwxrwxrwx  1 root root 0 Sep 24 22:36 dtpm -> ../../devices/virtual/powercap/dtpm
lrwxrwxrwx  1 root root 0 Sep 24 22:36 intel-rapl -> ../../devices/virtual/powercap/intel-rapl
lrwxrwxrwx  1 root root 0 Sep 24 22:36 intel-rapl:0 -> ../../devices/virtual/powercap/intel-rapl/intel-rapl:0
lrwxrwxrwx  1 root root 0 Sep 24 22:36 intel-rapl:0:0 -> ../../devices/virtual/powercap/intel-rapl/intel-rapl:0/intel-rapl:0:0
lrwxrwxrwx  1 root root 0 Sep 24 22:36 intel-rapl:1 -> ../../devices/virtual/powercap/intel-rapl/intel-rapl:1
lrwxrwxrwx  1 root root 0 Sep 24 22:36 intel-rapl:1:0 -> ../../devices/virtual/powercap/intel-rapl/intel-rapl:1/intel-rapl:1:0
```

Is your Kind cluster as `root`?

Reply:
Okay thats interesting, you have access to the powercap directory in your Kind Cluster without mounting it?  
The permissions of folder /sys/class/powercap look same. How about the file energy_uj inside of the intel directories? because thats where the problem was for me, as they couldnt be read. 

In /sys/class/powercap: 
total 0
drwxr-xr-x  2 root root 0 Okt  2 10:39 .
drwxr-xr-x 81 root root 0 Okt  2 10:39 ..
lrwxrwxrwx  1 root root 0 Okt  2 10:40 intel-rapl -> ../../devices/virtual/powercap/intel-rapl
lrwxrwxrwx  1 root root 0 Okt  2 10:40 intel-rapl:0 -> ../../devices/virtual/powercap/intel-rapl/intel-rapl:0
lrwxrwxrwx  1 root root 0 Okt  2 10:40 intel-rapl:0:0 -> ../../devices/virtual/powercap/intel-rapl/intel-rapl:0/intel-rapl:0:0
lrwxrwxrwx  1 root root 0 Okt  2 10:40 intel-rapl:0:1 -> ../../devices/virtual/powercap/intel-rapl/intel-rapl:0/intel-rapl:0:1
lrwxrwxrwx  1 root root 0 Okt  2 10:40 intel-rapl:0:2 -> ../../devices/virtual/powercap/intel-rapl/intel-rapl:0/intel-rapl:0:2
lrwxrwxrwx  1 root root 0 Okt  2 10:40 intel-rapl:1 -> ../../devices/virtual/powercap/intel-rapl/intel-rapl:1

The permissions of the energy_uj file inside /sys/class/powercap/intel-rapl:0 :
-r-------- 1 root root 4096 Okt  2 10:40 energy_uj

And yes the cluster, thus the docker container, runs as root

Reply:
chmod is not a good idea considering with security... as less as possible with access permission.

Reply:
```
sudo ls -la /sys/class/powercap/../../devices/virtual/powercap/intel-rapl/intel-rapl\:0/
total 0
drwxr-xr-x 4 root root    0 Sep 24 22:36 .
drwxr-xr-x 5 root root    0 Sep 24 22:36 ..
-r--r--r-- 1 root root 4096 Sep 24 22:36 constraint_0_max_power_uw
-r--r--r-- 1 root root 4096 Sep 24 22:36 constraint_0_name
-rw-r--r-- 1 root root 4096 Sep 24 22:36 constraint_0_power_limit_uw
-rw-r--r-- 1 root root 4096 Sep 24 22:36 constraint_0_time_window_us
-r--r--r-- 1 root root 4096 Sep 24 22:36 constraint_1_max_power_uw
-r--r--r-- 1 root root 4096 Sep 24 22:36 constraint_1_name
-rw-r--r-- 1 root root 4096 Sep 24 22:36 constraint_1_power_limit_uw
-rw-r--r-- 1 root root 4096 Sep 24 22:36 constraint_1_time_window_us
lrwxrwxrwx 1 root root    0 Sep 24 22:36 device -> ../../intel-rapl
-rw-r--r-- 1 root root 4096 Sep 24 22:36 enabled
-r-------- 1 root root 4096 Sep 24 22:36 energy_uj
drwxr-xr-x 3 root root    0 Sep 24 22:36 intel-rapl:0:0
-r--r--r-- 1 root root 4096 Sep 24 22:36 max_energy_range_uj
-r--r--r-- 1 root root 4096 Sep 24 22:36 name
drwxr-xr-x 2 root root    0 Sep 24 22:36 power
lrwxrwxrwx 1 root root    0 Sep 24 22:36 subsystem -> ../../../../../class/powercap
-rw-r--r-- 1 root root 4096 Sep 24 22:36 uevent
````

Reply:
My `energy_uj` is also read only.

Reply:
Did you run `make cluster-up` to build the Kind cluster?

Reply:
@marceloamaral I haven't before, but I tried it now and it does have the some effect, both ACPI and RAPL are not mounted/recognized although I have them on my local system, everything else such as eBPF is working. Could it be related with the log "getCPUArch failure"? :


I1004 18:36:28.303026       1 gpu.go:46] Failed to init nvml, err: could not init nvml: error opening libnvidia-ml.so.1: libnvidia-ml.so.1: cannot open shared object file: No such file or directory
**E1004 18:36:28.309454       1 utils.go:98] getCPUArch failure: no CPU power model found for architecture  Intel Skylake** 
I1004 18:36:28.312097       1 qat.go:35] Failed to init qat-telemtry err: could not get qat status exit status 127
I1004 18:36:28.322274       1 exporter.go:158] Kepler running on version: 02c16c1
I1004 18:36:28.322318       1 config.go:270] using gCgroup ID in the BPF program: true
I1004 18:36:28.322568       1 config.go:272] kernel version: 5.15
I1004 18:36:28.323485       1 exporter.go:170] LibbpfBuilt: true, BccBuilt: false
I1004 18:36:28.323531       1 exporter.go:189] EnabledBPFBatchDelete: true
I1004 18:36:28.324279       1 power.go:71] Unable to obtain power, use estimate method
I1004 18:36:28.324421       1 redfish.go:173] failed to initialize node credential: no supported node credential implementation
I1004 18:36:28.324440       1 power.go:56] use acpi to obtain power
I1004 18:36:28.325040       1 acpi.go:67] Could not find any ACPI power meter path. Is it a VM?
I1004 18:36:28.387118       1 exporter.go:204] Initializing the GPU collector
I1004 18:36:34.387593       1 watcher.go:66] Using in cluster k8s config

May I ask what OS and Kernel version you have? Also, in your Kind Cluster is it powercap that is being used and not msr?

Reply:
> E1004 18:36:28.309454 1 utils.go:98] getCPUArch failure: no CPU power model found for architecture Intel Skylake

This issue should have been fixed by Kepler PR#902. See latest code [here](https://github.com/sustainable-computing-io/kepler/blob/main/data/normalized_cpu_arch.csv#L6).
But I do not think it is related to your RAPL sysfs inaccessible issue.

Reply:
@emoek , it looks like your machine is a Intel Skylake BareMetal host, am I right?
So please confirm your test step first:
1. Run "make cluster-up" under kepler repo
2. Run "make build-manifest OPTS='CI_DEPLOY'"
3. Run "make cluster-deploy" to deploy kepler in Kind
4. Check Kepler log as above, or run "kubectl exec" command like this:
$ kubectl exec -i kepler-exporter-x9pmx -n kepler -- cat /sys/class/powercap/intel-rapl:0/energy_uj
232816879939

Reply:
> Okay thats interesting, you have access to the powercap directory in your Kind Cluster without mounting it? The permissions of folder /sys/class/powercap look same. How about the file energy_uj inside of the intel directories? because thats where the problem was for me, as they couldnt be read.
> 
> In /sys/class/powercap: total 0 drwxr-xr-x 2 root root 0 Okt 2 10:39 . drwxr-xr-x 81 root root 0 Okt 2 10:39 .. lrwxrwxrwx 1 root root 0 Okt 2 10:40 intel-rapl -> ../../devices/virtual/powercap/intel-rapl lrwxrwxrwx 1 root root 0 Okt 2 10:40 intel-rapl:0 -> ../../devices/virtual/powercap/intel-rapl/intel-rapl:0 lrwxrwxrwx 1 root root 0 Okt 2 10:40 intel-rapl:0:0 -> ../../devices/virtual/powercap/intel-rapl/intel-rapl:0/intel-rapl:0:0 lrwxrwxrwx 1 root root 0 Okt 2 10:40 intel-rapl:0:1 -> ../../devices/virtual/powercap/intel-rapl/intel-rapl:0/intel-rapl:0:1 lrwxrwxrwx 1 root root 0 Okt 2 10:40 intel-rapl:0:2 -> ../../devices/virtual/powercap/intel-rapl/intel-rapl:0/intel-rapl:0:2 lrwxrwxrwx 1 root root 0 Okt 2 10:40 intel-rapl:1 -> ../../devices/virtual/powercap/intel-rapl/intel-rapl:1
> 
> The permissions of the energy_uj file inside /sys/class/powercap/intel-rapl:0 : -r-------- 1 root root 4096 Okt 2 10:40 energy_uj
> 
> And yes the cluster, thus the docker container, runs as root

May I ask two questions?
1. Where did you run above commands, in which container?
2. Where did the Kind cluster run, VM hosted on BM or BM directly?